x-pack/plugins/fleet/server/routes/agent/upgrade_handler.ts

```diff
@@ -99,6 +100,16 @@ export const postAgentUpgradeHandler: RequestHandler<
         },
       });
     }
+
+    if (isAgentUpgrading(agent)) {
```

Comment:
I was tempted to refactor the `isAgentUpgradeable` function which is called below to avoid the redundancy checks (this one, `hasBeenUpgradedRecently` and unenrolled agents), but that would mean returning an error code and error message as well (not just a boolean), and this function is used in a few other spots. Let me know if you think it would be worth it.

Reply:
Do we have to use `isAgentUpgrading` and `isAgentUpgradeable` in different places? 

I'm wondering if we should still allow upgrade with force flag if agent is upgrading, e.g. if the upgrade details is stuck in a state other than failed due to a bug.
Eventually we should also update the condition of the restart upgrade action, where we determine that the agent is stuck in udpating if started 2 hours ago. This can be a separate issue.

Reply:
As is, `isAgentUpgrading` is used by `isAgentUpgradeable` and independently by this `postAgentUpgradeHandler`. The handler does call `isAgentUpgradeable` and return 400 with the generic error message `agent {id} is not upgradeable` if it fails. Before that, though, it performs other checks (upgraded less than 10 min ago, unenrolled, upgrading) and returns a specific status and error message if those fail. I thought it wasn't ideal that it's going to redo those checks when `isAgentUpgradeable` is called, hence why I thought this could be refactored.

As for `isAgentUpgradeable`, it is used in a good few places:
![Screenshot 2023-11-10 at 11 09 35](https://github.com/elastic/kibana/assets/23701614/d1bfc4f8-c4cd-4e7d-a191-8a160a63bb0b)

So we could presumable refactor it to return more than just a boolean, but these would need to be amended accordingly.

Reply:
> I'm wondering if we should still allow upgrade with force flag if agent is upgrading, e.g. if the upgrade details is stuck in a state other than failed due to a bug.

Great point! We prevent it when the upgrade cooldown is not over, but you're right that it should work in this case. I'll check that.

> Eventually we should also update the condition of the restart upgrade action, where we determine that the agent is stuck in udpating if started 2 hours ago, we should use the upgrade details. This can be a separate issue.

Also a great point, thank you 👍 I can file the issue if it doesn't exist already.

Reply:
hm I'm not sure if it's worth refactoring now, it seems more intuitive to keep returning a boolean from functions called `isX`

Reply:
I agree, that's why I decided against the refactor. It doesn't seem like expensive operations anyway.